### PR TITLE
Added weight compression algorithm for OpenVINO backend

### PR DESCRIPTION
**Pull request by l-bat**
### Changes

Extended data free int8 weight compression algorithm for OpenVINO backend

Example (WeightsModel):
image

PR to optimum https://github.com/huggingface/optimum-intel/pull/415

### Reason for changes

Optimize the model footprint and performance of large models where the size of weights is relatively larger than the size of activations

### Related tickets

117412

### Tests

`tests/openvino/native/quantization/test_weights_compression.py`
swin transformer support verified

Results
Task: lambada_openai
|     Model |Metric|Value |   |Stderr|
|--------------|------|-----:|---|-----:|
|dolly-v2-3b_original| ppl   |5.0144|±  |0.1510|
|              |acc   |0.6297|±  |0.0067|
|dolly-v2-3b_compressed|ppl   |4.9868|±  |0.1498|
|                |acc  |0.6313|±  |0.0067|
|Llama-2-7b-chat-hf_original|ppl   |3.2788|±  |0.0866|
|       |acc   |0.7058|±  |0.0063|
|Llama-2-7b-chat-hf_compressed|ppl   |3.2856|±  |0.0869|
|       |acc   |0.7054|±  |0.0064|

### CODE REVIEW

File: nncf/quantization/quantize_model.py [hunk 1]
```diff
@@ -226,22 +226,22 @@ def quantize_with_accuracy_control(
 
 
 @api(canonical_alias="nncf.compress_weights")
-def compress_weights(model: TModel, use_fake_quantize: bool = False) -> TModel:
+def compress_weights(model: TModel) -> TModel:
     """
     Compress model weights.
```
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply nncf/quantization/quantize_model.py:231]: It is not an API-breaking change, since the `compress_weights` functionality was introduced in this release.
See https://github.com/openvinotoolkit/nncf/pull/1941 for details.

File: nncf/openvino/quantization/weights_compression.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            input_high = np.max(weight, axis=axes, keepdims=True)
+            stats = OVMinMaxTensorStatistic(input_low, input_high)
+            fq_params = get_fq_params(stats)
+
+            node_name = node.get_friendly_name()
+            fq_name = f"{node_name}/fq_weights_{const_port_id}"
+            _insert_fake_quantize(fq_params, weight_output, fq_name)
+
+
+def _get_reduction_axes(metatype: Type[OperatorMetatype], node: ov.Node, const_port_id: int) -> Union[int, Tuple[int]]:
```
Comment (KodiaqQ) [inline nncf/openvino/quantization/weights_compression.py:92]: Don't we need this method in the `node_utils.py` to reuse in the `MinMaxQuantization` backend? Or is it not the same?

File: nncf/openvino/quantization/weights_compression.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        channel_axes = list(range(ndims - 2))
+        if target_dim < ndims:
+            channel_axes.append(target_dim)
+        axes = tuple(i for i in range(ndims) if i not in channel_axes)
+    else:
+        axes = 1
+    return axes
+
+
+def _insert_fake_quantize(fq_params: FakeQuantizeParameters, weight_output: ov.Output, fq_name: str):
```
Comment (KodiaqQ) [inline nncf/openvino/quantization/weights_compression.py:116]: Can't we reuse `ModelTransformer` here?

File: nncf/openvino/quantization/quantize_model.py [hunk 4]
```diff
@@ -369,3 +370,13 @@ def quantize_with_accuracy_control_impl(
         advanced_quantization_parameters,
         advanced_accuracy_restorer_parameters,
     )
+
+
+def compress_weights_impl(model: ov.Model) -> ov.Model:
+    """
+    Implementation of the `compress_weights()` method for the OpenVINO backend.
+    """
+    insert_pre_compression_operations(model)
```
[the remark this answers is not in the document]
Comment (l-bat) [reply nncf/openvino/quantization/quantize_model.py:379]: The compression applies in-place according to https://github.com/openvinotoolkit/nncf/blob/develop/docs/compression_algorithms/CompressWeights.md

File: nncf/openvino/quantization/quantize_model.py [hunk 5]
```diff
@@ -369,3 +370,13 @@ def quantize_with_accuracy_control_impl(
         advanced_quantization_parameters,
         advanced_accuracy_restorer_parameters,
     )
+
+
+def compress_weights_impl(model: ov.Model) -> ov.Model:
+    """
+    Implementation of the `compress_weights()` method for the OpenVINO backend.
+    """
```
[the remark this answers is not in the document]
Comment (l-bat) [reply nncf/openvino/quantization/quantize_model.py:378]: Description is similar to `native_quantize_impl` and `native_quantize_with_accuracy_control_impl`

File: nncf/openvino/quantization/weights_compression.py (hunk 2 above)
Comment (l-bat) [reply nncf/openvino/quantization/weights_compression.py:93]: In `MinMaxQuantization` we use `NNCFGraph`. 
We don't create `NNCFGraph` from OpenVINO Model for `WeightCompression` to speed up algorithm.
I added `get_matmul_channel_axes` to `node_utils.py`

**Comment by l-bat** [comment]
run pytorch pre-commit tests
run openvino pre-commit tests

File: nncf/openvino/quantization/weights_compression.py (hunk 3 above)
Comment (l-bat) [reply nncf/openvino/quantization/weights_compression.py]: Removed

File: nncf/openvino/quantization/weights_compression.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    :return: The reduction axes as an integer or a tuple of integers.
+    """
+    if metatype is OVMatMulMetatype:
+        transpose = node.get_attributes()[f"transpose_{'a' if weight_port_id == 0 else 'b'}"]
+        ndims = node.input(weight_port_id).get_partial_shape().rank.get_max_length()
+        channel_axes = get_matmul_channel_axes(weight_port_id, ndims, transpose)
+        axes = tuple(i for i in range(ndims) if i not in channel_axes)
+    elif metatype is OVEmbeddingMetatype:
+        axes = (metatype.const_channel_axis[0] + 1) % 2
```
Comment (KodiaqQ) [inline nncf/openvino/quantization/weights_compression.py:122]: Could you explain these calculations?

Comment (l-bat) [reply nncf/openvino/quantization/weights_compression.py:122]: To align with torch https://github.com/openvinotoolkit/nncf/blob/c027c8b43c4865d46b8de01d8350dd338ec5a874/nncf/torch/quantization/weights_compression.py#L82-L83

File: tests/openvino/native/quantization/test_weights_compression.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+def test_compress_weights(model_creator_func):
+    ref_compressed_weights = TEST_MODELS[model_creator_func]
+    model = model_creator_func().ov_model
+    compressed_model = compress_weights(model)
+
+    n_compressed_weights = 0
+    for op in compressed_model.get_ops():
+        if op.get_type_name() == "Constant" and op.get_friendly_name() in ref_compressed_weights:
+            assert op.get_element_type() == ov.Type(np.int8)
+            n_compressed_weights += 1
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_weights_compression.py:36]: I'm afraid this kind of check is not enough to verify that quantization is done correctly.
Could you add a very basic model with ~2-10 weights and calculate expected output?
Or you could catch only quantized weights and compare with the reference. 
The current state of the branch passes the graph test, but accuracy is around 0 on the real model.

Comment (ljaljushkin) [reply tests/openvino/native/quantization/test_weights_compression.py:36]: Sorry, I incorrectly verified accuracy. Seems like there's no issue with accuracy on the current branch. But still, I'd think about better verification than graph test.

Comment (l-bat) [reply tests/openvino/native/quantization/test_weights_compression.py:40]: Added `test_compare_compressed_weights`

File: tests/openvino/native/quantization/test_weights_compression.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    for op in compressed_model.get_ops():
+        if op.get_type_name() == "Constant" and op.get_friendly_name() in ref_compressed_weights:
+            assert op.get_element_type() == ov.Type(np.uint8)
+            n_compressed_weights += 1
+
+    assert n_compressed_weights == len(ref_compressed_weights)
+
+
+def test_compare_compressed_weights():
+    model = IntegerModel().ov_model
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_weights_compression.py:46]: Isn't it too big for comparing stats? What about creating smaller model in order to make reference file smaller?

File: tests/openvino/native/quantization/test_weights_compression.py (hunk 7 above)
Comment (ljaljushkin) [reply tests/openvino/native/quantization/test_weights_compression.py:40]: Thanks!

File: nncf/openvino/quantization/weights_compression.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            original_weight_dtype = weight_output.get_element_type().to_dtype()
+            if original_weight_dtype not in [np.float32, np.float16]:
+                continue
+
+            weight = get_const_value(weight_node)
+            axes = _get_reduction_axes(metatype, node, const_port_id)
+            min_values = np.min(weight, axis=axes, keepdims=True)
+            max_values = np.max(weight, axis=axes, keepdims=True)
+
+            scale = (max_values - min_values) / (level_high - level_low)
```
Comment (kshpv) [inline nncf/openvino/quantization/weights_compression.py:63]: Do not you want to reuse `calculate_quantizer_parameters()` from `nncf/quantization/fake_quantize.py`?

File: nncf/openvino/quantization/weights_compression.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    :param model: The model to be transformed.
+    :param bits: Number of bits for quantization.
+    """
+    allowed_metatypes_to_const_port = {OVEmbeddingMetatype: [0], OVMatMulMetatype: [0, 1]}
+    level_low = 0
+    level_high = 2**bits - 1
+
+    for node in model.get_ops():
+        # pylint:disable=protected-access
+        metatype = GraphConverter._get_node_metatype(node)
```
Comment (kshpv) [inline nncf/openvino/quantization/weights_compression.py:41]: I would suggest to avoid `# pylint:disable=protected-access` by renaming this function to `get_node_metatype()`. Or maybe the better approach is to have this function inside `nncf/nncf/openvino/graph/metatypes/openvino_metatypes.py`

File: nncf/openvino/quantization/weights_compression.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            weight_node = get_operation_const_op(node, const_port_id)
+            if weight_node is None:
+                continue
+
+            weight_output = weight_node.output(0)
+            weight_name = weight_node.get_friendly_name()
+            target_inputs = weight_output.get_target_inputs()
+
+            original_weight_dtype = weight_output.get_element_type().to_dtype()
+            if original_weight_dtype not in [np.float32, np.float16]:
```
Comment (kshpv) [inline nncf/openvino/quantization/weights_compression.py:55]: Can it be `float64`? https://github.com/openvinotoolkit/nncf/blob/develop/nncf/openvino/graph/nncf_graph_builder.py#L48

File: nncf/openvino/quantization/weights_compression.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                continue
+
+            weight = get_const_value(weight_node)
+            axes = _get_reduction_axes(metatype, node, const_port_id)
+            min_values = np.min(weight, axis=axes, keepdims=True)
```
Comment (kshpv) [inline nncf/openvino/quantization/weights_compression.py:60]: Do we need to support not only `min` and `max` calculation for weights?

**Review by kshpv — CHANGES_REQUESTED** [review]

File: nncf/openvino/quantization/weights_compression.py (hunk 9 above)
Comment (kshpv) [inline nncf/openvino/quantization/weights_compression.py:63]: Do you want to reuse `calculate_scale_zero_point()` from `nncf/nncf/onnx/quantization/quantizer_parameters.py` ?

**Review by kshpv — CHANGES_REQUESTED** [review]

File: nncf/openvino/quantization/weights_compression.py (hunk 11 above)
Comment (l-bat) [reply nncf/openvino/quantization/weights_compression.py:55]: I'm not sure that models can have `Embedding` or ``MatMul` weights in `float64`

File: nncf/openvino/quantization/weights_compression.py (hunk 12 above)
Comment (l-bat) [reply nncf/openvino/quantization/weights_compression.py:60]: Not for this release

**Review by KodiaqQ — APPROVED** [review]

**Review by kshpv — APPROVED** [review]

**Review by vshampor — APPROVED** [review]

**Comment by l-bat** [comment]
run tensorflow pre-commit tests

**Review by andreyanufr — APPROVED** [review]

File: nncf/openvino/quantization/weights_compression.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            weight_output = weight_node.output(0)
+            weight_name = weight_node.get_friendly_name()
+            target_inputs = weight_output.get_target_inputs()
+
+            original_weight_dtype = weight_output.get_element_type().to_dtype()
+            if original_weight_dtype not in [np.float32, np.float16, np.float64]:
+                continue
+
+            weight = get_const_value(weight_node)
+            axes = _get_reduction_axes(metatype, node, const_port_id)
```
Comment (alexsu52) [inline nncf/openvino/quantization/weights_compression.py:59]: Does it necessary to handle reshape here https://github.com/openvinotoolkit/nncf/blob/2f631fc28db634280c421ccd3b172509ac4e65c0/nncf/openvino/graph/metatypes/openvino_metatypes.py#L719?

**Review by alexsu52 — COMMENTED** [review]
LGTM

File: nncf/quantization/quantize_model.py [hunk 14]
```diff
@@ -226,22 +226,22 @@ def quantize_with_accuracy_control(
...
     backend = get_backend(model)
     if backend == BackendType.TORCH:
         import nncf.torch
 
-        return nncf.torch.compress_weights(model, use_fake_quantize)
+        return nncf.torch.compress_weights(model)
+    if backend == BackendType.OPENVINO:
+        from nncf.openvino.quantization.quantize_model import compress_weights_impl
+
+        return compress_weights_impl(model)
```
[the remark this answers is not in the document]
Comment (l-bat) [reply nncf/quantization/quantize_model.py:244]: The code looks consistent with `quantize_impl` and `quantize_with_accuracy_control_impl`

File: nncf/openvino/quantization/weights_compression.py (hunk 13 above)
Comment (alexsu52) [reply nncf/openvino/quantization/weights_compression.py:59]: Are you going to remove support of # (Constant) -> (Convert) -> (FakeQuantize) -> (Reshape) -> (Operation)  in the follow-up PR?

**Review by alexsu52 — APPROVED** [review]
